trieste/pareto.py

```diff
@@ -0,0 +1,132 @@
+        P is the number of points in pareto front
+        dominances gives the number of dominating points for each data point
+
+
+    """
+    extended = tf.tile(observations[None], [len(observations), 1, 1])
+    swapped_ext = tf.transpose(extended, [1, 0, 2])
+    dominance = tf.math.count_nonzero(
+        tf.logical_and(
+            tf.reduce_all(extended <= swapped_ext, axis=2),
+            tf.reduce_any(extended < swapped_ext, axis=2),
+        ),
+        axis=1,
+    )
+
+    return tf.boolean_mask(observations, dominance == 0), dominance
+
+
+class BoundedVolumes:
+    """
+    A :class:`BoundedVolumes store the index of the Pareto front to form lower and upper
+    bounds of the pseudo cells decomposition.
+    """
+
+    def __init__(self, lower_idx: tf.Tensor, upper_idx: tf.Tensor):
+        """
+        Construct bounded volumes.
+
+        :param lower_idx: the lowerbounds index of the volumes
+        :param upper_idx: the upperbounds index of the volumes
+        """
+
+        assert tf.reduce_all(lower_idx.shape == upper_idx.shape)
+        self.lower_idx = lower_idx
+        self.upper_idx = upper_idx
+
+    def append(self, lower_idx: tf.Tensor, upper_idx: tf.Tensor):
+        """
+        Add new bounded volumes.
+
+        :param lower_idx: the lowerbounds of the volumes
+        :param upper_idx: the upperbounds of the volumes
+        """
+        self.lower_idx = tf.concat([self.lower_idx, lower_idx], axis=0)
+        self.upper_idx = tf.concat([self.upper_idx, upper_idx], axis=0)
+
+    def clear(self):
+        # Clears all stored bounded volumes
+        dtype = self.lower_idx.dtype
+        outdim = self.lower_idx.shape[1]
+        self.lower_idx = tf.zeros([0, outdim], dtype=dtype)
+        self.upper_idx = tf.zeros([0, outdim], dtype=dtype)
+
+    def size(self) -> tf.Tensor:
+        """
+        :return: volume of each bounded volume
+        """
+        return tf.math.reduce_prod(self.upper_idx - self.lower_idx, axis=0, keepdims=True)
+
+
+class Pareto:
+    """
+    A :class:`Pareto` Construct a Pareto set.
+    Stores a Pareto set and calculates the cell bounds covering the non-dominated region.
+    The latter is needed for certain multiobjective acquisition functions.
+    """
+
+    def __init__(self, observations: TensorType):
+        """
+        :param observations: The observations for all objectives, with shape [N, 2].
+        :raise ValueError (or InvalidArgumentError): If ``observations`` has an invalid shape.
+        """
+        tf.debugging.assert_shapes([(observations, [None, 2])])
+
+        pf, _ = non_dominated(observations)
+        self.front: Final[TensorType] = tf.gather_nd(pf, tf.argsort(pf[:, :1], axis=0))
+        self.bounds: Final[BoundedVolumes] = self._bounds_2d(self.front)
+
+    @staticmethod
+    def _bounds_2d(front: TensorType) -> BoundedVolumes:
+        """
+        :param front: pareto front of current observations
+        """
```

Comment:
adding a docstring will make the method show up in the docs. We don't want that here do we?

Reply:
removed